src/intent_parser/lookup_intent_parser.rs

```diff
@@ -0,0 +1,801 @@
+        let b = self
+            .shared_resources
+            .builtin_entity_parser
+            .extract_entities(input, None, true)?
+            .into_iter()
+            .map(|entity| entity.into());
+        // get custom entities
+        let c = self
+            .shared_resources
+            .custom_entity_parser
+            .extract_entities(input, None)?
+            .into_iter()
+            .map(|entity| entity.into());
+
+        // combine entities
+        let mut all_entities: Vec<MatchedEntity> = Vec::new();
+        all_entities.extend(b);
+        all_entities.extend(c);
+
+        Ok(deduplicate_overlapping_entities(all_entities))
+    }
+
+    fn parse_map_output(
+        &self,
+        input: &str,
+        output: Option<&(i32, Vec<i32>)>,
+        entities: Vec<MatchedEntity>,
+        intents: Option<&[&str]>,
+    ) -> InternalParsingResult {
+        if let Some((intent_id, slots_ids)) = output {
+            // assert invariant: length of slot ids matches that of entities
+            debug_assert!(slots_ids.len() == entities.len());
+            let intent_name = &self.intents_names[*intent_id as usize];
+            if intents
+                .unwrap_or(&[intent_name.as_ref()])
+                .contains(&intent_name.as_ref())
+            {
+                let intent = IntentClassifierResult {
+                    intent_name: Some(intent_name.to_string()),
+                    confidence_score: 1.0,
+                };
+                // get slots and return result
+                // we assume entities are sorted by their ranges
+                let mut slots = vec![];
+                for (slot_id, entity) in slots_ids.iter().zip(entities.iter()) {
+                    let slot_name = &self.slots_names[*slot_id as usize];
+                    let entity_name = &entity.entity_name;
+                    let char_range = &entity.range;
+                    let value = substring_with_char_range(input.to_string(), &char_range);
+                    slots.push(InternalSlot {
+                        value,
+                        char_range: char_range.clone(),
+                        entity: entity_name.to_string(),
+                        slot_name: slot_name.to_string(),
+                    });
+                }
+                InternalParsingResult { intent, slots }
+            } else {
+                // if intent name not in intents, return empty result
+                InternalParsingResult::empty()
+            }
+        } else {
+            // if lookup value was none return empty result
+            InternalParsingResult::empty()
+        }
+    }
+}
+
+impl IntentParser for LookupIntentParser {
+    fn parse(
+        &self,
+        input: &str,
+        intents_whitelist: Option<&[&str]>,
+    ) -> Result<InternalParsingResult> {
+        let mut entities = self.get_all_entities(input)?;
+        let formatted_input = LookupIntentParser::replace_entities(input, &mut entities);
```

Comment:
Don't you think `replace_entities` should rather be a free function ?

Reply:
I don't see the logic being used anywhere beyond the lookup intent parser

Reply:
I agree, but it's rather that this function is more a utils than an associated function like a specific constructor.